src/main/java/com/mongodb/kafka/connect/sink/StartedMongoSinkTask.java

```diff
@@ -164,107 +159,19 @@ private void handleTolerableWriteException(
-        assertNotNull(writeConcernError);
-        for (int i = 0; i < batch.size(); i++) {
-          recordsFailedWithWriteConcernError.put(i, batch.get(i));
-        }
-      } else {
-        assertTrue(writeErrors.size() == 1);
-        BulkWriteError writeError = writeErrors.get(0);
-        int writeErrorIdx = writeError.getIndex();
-        recordsFailedWithWriteError.put(
-            writeErrorIdx,
-            new SimpleImmutableEntry<>(batch.get(writeErrorIdx), new WriteException(writeError)));
-        for (int i = writeErrorIdx + 1; i < batch.size(); i++) {
-          skippedRecords.put(i, batch.get(i));
-        }
-        if (writeConcernError != null) {
-          for (int i = 0; i < writeErrorIdx; i++) {
-            recordsFailedWithWriteConcernError.put(i, batch.get(i));
-          }
-        }
-      }
-    }
-
-    void log() {
-      if (!recordsFailedWithWriteError.isEmpty()) {
-        recordsFailedWithWriteError.forEach(
-            (i, recordFailedWithWriteError) ->
-                StartedMongoSinkTask.log(
-                    singletonList(recordFailedWithWriteError.getKey()),
-                    recordFailedWithWriteError.getValue()));
-      }
-      if (!recordsFailedWithWriteConcernError.isEmpty()) {
-        StartedMongoSinkTask.log(
-            recordsFailedWithWriteConcernError.values(), assertNotNull(writeConcernException));
-      }
-      if (!skippedRecords.isEmpty()) {
-        StartedMongoSinkTask.log(skippedRecords.values(), writeSkippedException);
-      }
-      LOGGER.error(null, e);
-    }
-
-    /**
-     * {@linkplain #errorReporter Reports} records in the batch that either definitely have not been
-     * written, or may not have been written. While it is unclear whether the order of reported records in the DLQ is the same as the order
-     * in which they are reported, we still report them in the order they are present in the batch just in case because it is trivial to do.
-     */
-    void report() {
-      for (int i = 0; i < batch.size(); i++) {
-        Entry<SinkRecord, WriteException> recordFailedWithWriteError =
-            recordsFailedWithWriteError.get(i);
-        if (recordFailedWithWriteError != null) {
-          errorReporter.accept(
-              recordFailedWithWriteError.getKey(), recordFailedWithWriteError.getValue());
-          continue;
-        }
-        SinkRecord recordFailedWithWriteConcernError = recordsFailedWithWriteConcernError.get(i);
-        if (recordFailedWithWriteConcernError != null) {
-          errorReporter.accept(
-              recordFailedWithWriteConcernError, assertNotNull(writeConcernException));
-          continue;
-        }
-        SinkRecord skippedRecord = skippedRecords.get(i);
-        if (skippedRecord != null) {
-          errorReporter.accept(skippedRecord, writeSkippedException);
-        }
-      }
-    }
+  public static void log(final Collection<SinkRecord> records, final RuntimeException e) {
```

Comment:
If you opt to decouple the logging mechanism and inject it instead this could be:

```
private static BiConsumer<Collection<SinkRecord>, RuntimeException>  ERROR_LOGGER =
          (records, e) -> LOGGER.error("Failed to put into the sink the following records: {}", records, e);```

Reply:
I decoupled `AnalyzedBatchFailedWithBulkWriteException` from `StartedMongoSinkTask`, but replacing this method with an object does not seem beneficial for the code nonetheless. Take a look, maybe you'll agree.

Reply:
Ack